### ISSUE TITLE
Improve drawing of text model

### ISSUE THREAD

**Issue by douwevos** [douwevos/natpad3#17]
Currently all lines in the text-model are drawn on offscreen images in the method invalidateLines. This is fine for very small documents but will take a huge amount of memory (offscreen images) for large smaller documents and will be very slow. The goal is to only draw lines which changed and are visible. The View class will need an additional field m_view_height and the invalidateLines should now only create/update lines which are between m_view_y and m_view_y+m_view_height.

A second optimization is that instead of always creating new offscreen images and drawing lines the invalidateLines should only update lines that changed. The List of offsreen images passed to the draw method therefor must be replaced with a list of a new object. The new object (name maybe CachedLine?) holds besides the offscreen image also a reference to line. The invalidateLines should now search for each line that is between m_view_y and m_view_y+m_view_height for that specific line in the list of CachedLine instances. Please not that its easier to create a new list of CachedLine instance each time invalidateLines is called instead of manipulating the old list.

References in the current implementation (DocumentView.c == View.cpp). It may help with implementing it :)

The place where the cached-lines are stored in the DocumentView
https://github.com/douwevos/natpad/blob/be807b06508ab2293c07f6233fb332a57939e2d7/chameleon/src/layout/chadocumentview.c#L115

A new cache-list is created in invalidate-lines
https://github.com/douwevos/natpad/blob/be807b06508ab2293c07f6233fb332a57939e2d7/chameleon/src/layout/chadocumentview.c#L1923

Searching for an old version of a CachedLine (ChaSubLineCache)
https://github.com/douwevos/natpad/blob/be807b06508ab2293c07f6233fb332a57939e2d7/chameleon/src/layout/chadocumentview.c#L2218

Replacing the old cache-list with the new one
https://github.com/douwevos/natpad/blob/be807b06508ab2293c07f6233fb332a57939e2d7/chameleon/src/layout/chadocumentview.c#L2295

**Comment by gerardvisser** [comment]
What about when the window is resized? Then the on_draw method is also called.

**Comment by douwevos** [comment]
> What about when the window is resized? Then the on_draw method is also called.

Yes this is true. This is a special case where there are 5 possible scenarios (focusing on the width):

1. The new width is smaller. The editor is not wrapping lines.
2. The new width is larger. The editor is not wrapping lines
3. The new width is smaller. The editor is wrapping lines.
4. The new width is larger. The editor is wrapping lines.
5. The width did not change (only the height changed).

Scenario 1 would be the only scenario where you theoretically would not need to create new images.

The resizing of the editor window will be handled in a separate issue.

**Comment by douwevos** [comment]
https://github.com/douwevos/natpad/blob/be807b06508ab2293c07f6233fb332a57939e2d7/chameleon/src/chaeditor.c#L127

**Comment by gerardvisser** [comment]
I've implemented a View::invalidateLines that only creates images for those lines that are visible
between View::m_view_y and View::m_view_y + View::m_viewHeight.
The View::draw draws the lines that it finds in View::m_lineImages. So it copies the lineImages to the
Cairo::Context that represents the screen. These lines cover the whole screen if there are that many
lines left from the top.
The first paragraph in the issue description also talks about 'drawing only those lines that have changed'.
This, of course, introduces the problem of identifying these.
For now let's say that we did. Let's also assume that no scroll took place (m_view_y is equal to what it
was previously). Suppose that there are currently 25 lines visible. Now the goal of drawing only the
changed lines could mean:
1. View::invalidateLines fills the m_lineImages only with those lines that changed, which might be less
than the 25 visible. View::draw has to find out where to put these images on the screen.
2. View::invalidateLines fills m_lineImages with all 25 lines that are visible, but reuses the old images
of the lines that did not change. In this way only the changed lines are actually drawn, but View::draw
copies all 25 images to the screen anyway.
Question: which is the preferred way?
The second proposal is easier to implement, also if we take scrolling into account again (in which
case lines that did not change their text, did change their position on the screen).
The second paragraph of the issue (A second optimization is that .... ) seems to suggest this solution
by talking (indirectly) about reusing old images.
Let's come back to the problem of identifying those lines that changed. The only way to do this seems
to be by comparing the text for which we want to create an image, with the texts that were used to
create the old images. Whether the lines that are the same in this way are also the same to the user
or, on the other hand, two different lines that happen to have the same text, is unimportant for our
purpose. This seems to suggest, as does the description of the issue, that we add a copy of a reference
to the text to the object (an instance of LineImage) that holds the Cairo::Surface on which this text
was drawn.
Note that the class LineImage that I created for a previous issue, seems to fulfill the role of the
CachedLine class that you suggest in the issue.
